shapely/strtree.py

```diff
@@ -66,78 +68,94 @@ class STRtree:
+    Packing". https://ia600900.us.archive.org/27/items/nasa_techdoc_19970016975/19970016975.pdf
 
-    >>> tree = STRtree([])
-    >>> tree.query(Point(0, 0))
-    []
-    >>> print(tree.nearest(Point(0, 0)))
-    None
     """
 
-    def __init__(self, geoms):
+    def __init__(self, initdata):
         warn(
-            "STRtree will be completely changed in 2.0.0. The exact API is not yet decided, but will be documented before 1.8.0",
+            "STRtree will be changed in 2.0.0. The exact API is not yet decided, but will be documented before 1.8.0",
             ShapelyDeprecationWarning,
             stacklevel=2,
         )
-        # filter empty geometries out of the input
-        geoms = [geom for geom in geoms if not geom.is_empty]
-        self._n_geoms = len(geoms)
-
-        self._init_tree_handle(geoms)
-
-        # Keep references to geoms.
-        self._geoms = list(geoms)
-
-    def _init_tree_handle(self, geoms):
-        node_capacity = 10
-        self._tree_handle = lgeos.GEOSSTRtree_create(node_capacity)
-        for geom in geoms:
-            lgeos.GEOSSTRtree_insert(self._tree_handle, geom._geom, ctypes.py_object(geom))
+        self._tree = None
+        self._rev = None
+        if initdata:
+            self._rev = {
+                idx: geom
+                for geom, idx in self._iterinitdata(initdata)
+                if not geom.is_empty
+            }
+            if self._rev:
+                self._init_tree(self._rev.items())
+
+    def _iterinitdata(self, initdata):
+        for enum_idx, item in enumerate(initdata):
+            if isinstance(item, tuple):  # a geom, idx pair
+                yield item[:2]
+            else:  # a single geometry
+                yield (item, enum_idx)
+
+    def _init_tree(self, rev_initdata):
+        if rev_initdata:
+            node_capacity = 10
+            self._tree = lgeos.GEOSSTRtree_create(node_capacity)
+            for idx, geom in rev_initdata:
+                lgeos.GEOSSTRtree_insert(self._tree, geom._geom, ctypes.py_object(idx))
 
     def __getstate__(self):
         state = self.__dict__.copy()
-        del state["_tree_handle"]
+        del state["_tree"]
         return state
 
     def __setstate__(self, state):
         self.__dict__.update(state)
-        self._init_tree_handle(self._geoms)
+        if self._rev:
+            self._init_tree(self._rev.items())
 
     def __del__(self):
-        if self._tree_handle is not None:
+        if self._tree is not None:
             try:
-                lgeos.GEOSSTRtree_destroy(self._tree_handle)
+                lgeos.GEOSSTRtree_destroy(self._tree)
             except AttributeError:
                 pass  # lgeos might be empty on shutdown.
 
-            self._tree_handle = None
+            self._tree = None
 
-    def query(self, geom):
-        """
-        Search the index for geometry objects whose extents
-        intersect the extent of the given object.
+    def query_item(self, geom):
```

Comment:
Naming/terminology question: would eg "ids" be clearer as "items"? 
If you see "(geom, item)" it's of course clear that "item" is the identifier. But if I read "item" by itself (as in this function name / docstring), for me it would not directly be clear whether it references the geometry item of the id item ..

Reply:
"item" is the term GEOS uses. An item is stored in a node of the tree. Whether the item is an "id" or not is up to the application.